src/_pytest/faulthandler.py

```diff
@@ -17,70 +17,92 @@ def pytest_addoption(parser):
-        return sys.__stderr__.fileno()
+    if not faulthandler.is_enabled():
+        # faulthhandler is not enabled, so install plugin that does the actual work
+        # of enabling faulthandler before each test executes
+        config.pluginmanager.register(FaultHandlerHooks(), "faulthandler-hooks")
+    else:
+        from _pytest.warnings import _issue_warning_captured
+
+        # do not handle dumping to stderr if faulthandler is already enabled, so warn
+        # users that the option is being ignored
+        timeout = FaultHandlerHooks.get_timeout_config_value(config)
+        if timeout > 0:
+            _issue_warning_captured(
+                pytest.PytestConfigWarning(
+                    "faulthandler module enabled before pytest configuration step, "
+                    "'faulthandler_timeout' option ignored"
+                ),
+                config.hook,
+                stacklevel=2,
+            )
+
+
+class FaultHandlerHooks:
+    """Implements hooks that will actually install fault handler before tests execute,
+    as well as correctly handle pdb and internal errors."""
+
+    def pytest_configure(self, config):
+        import faulthandler
 
+        stderr_fd_copy = os.dup(self._get_stderr_fileno())
+        config.fault_handler_stderr = os.fdopen(stderr_fd_copy, "w")
+        faulthandler.enable(file=config.fault_handler_stderr)
 
-def pytest_unconfigure(config):
-    import faulthandler
+    def pytest_unconfigure(self, config):
+        import faulthandler
 
-    faulthandler.disable()
-    # close our dup file installed during pytest_configure
-    f = getattr(config, "fault_handler_stderr", None)
-    if f is not None:
+        faulthandler.disable()
+        # close our dup file installed during pytest_configure
         # re-enable the faulthandler, attaching it to the default sys.stderr
         # so we can see crashes after pytest has finished, usually during
         # garbage collection during interpreter shutdown
         config.fault_handler_stderr.close()
         del config.fault_handler_stderr
-        faulthandler.enable(file=_get_stderr_fileno())
-
-
-@pytest.hookimpl(hookwrapper=True)
-def pytest_runtest_protocol(item):
-    timeout = float(item.config.getini("faulthandler_timeout") or 0.0)
-    if timeout > 0:
-        import faulthandler
+        faulthandler.enable(file=self._get_stderr_fileno())
 
-        stderr = item.config.fault_handler_stderr
-        faulthandler.dump_traceback_later(timeout, file=stderr)
+    @staticmethod
+    def _get_stderr_fileno():
         try:
+            return sys.stderr.fileno()
+        except (AttributeError, io.UnsupportedOperation):
+            # python-xdist monkeypatches sys.stderr with an object that is not an actual file.
+            # https://docs.python.org/3/library/faulthandler.html#issue-with-file-descriptors
+            # This is potentially dangerous, but the best we can do.
+            return sys.__stderr__.fileno()
```

Comment:
Note: this is not covered (but also before).
Wonder if pytest-xdist has been fixed in this regard?
(Anyway, out of scope for this PR to cover it.)

Reply:
`pytest-xdist` can't really fix it while using `execnet` as backend, I'm afraid.

This code is covered in pytest-xdist's own suite, this can be covered using some mocking perhaps?

Reply:
I've thought we should have seen it being covered since we're testing with xdist also?

Reply:
Hmm yeah that is strange.

OK I found the PR which changed that to a try/except:

https://github.com/pytest-dev/pytest-faulthandler/commit/26cd5308ed41be2f800b0c0504e2e46baeb987ce

> If you try to test a pytest plugin using the pytester module, and you
also have pytest-faulthandler installed, then currently it crashes
with obscure errors. The problem is that by default, pytester runs the
sub-pytest in-process, and the sub-pytest has sys.stderr set to a
_pytest.capture.CaptureIO object, which has a fileno method (so the
old hasattr(sys.stderr, "fileno") check passes), but if you try to
call this method then it raises io.UnsupportedOperation.

> So this commit robustifies the sys.stderr.fileno() calling code to
handle both a missing fileno (as before), and it raising
UnsupportedOperation.

So it seems this is not really required anymore. I will push a change to remove this workaround, let's see.

Reply:
> So it seems this is not really required anymore. I will push a change to remove this workaround, let's see.

Better do it separately (in another PR against features later).

I think it is good to keep it in general (since there might be other "bad citizens" in this regard, but covering/triggering it would be nice of course).

Reply:
But the commit message gives more info, and it might not be required then really anymore:

> this change is no longer required now that `CaptureIO` subclasses
`io.TextIOWrapper` directly and we don't support Python 2.

However I'd still be conservative and do it in a separate PR (against features then likely).

Reply:
Fair enough, let's wait for the CI to finish and then I will do a separate PR.

Reply:
Will need to wait for the next merge of `master` into `features`. Keeping the hash here for reference so I can do it later (fcbfd89c30103ec3e871dec0e0c749ac55f7f6f8).